### ISSUE TITLE
Allow variable IDs in `prefer-query-selector`

### ISSUE THREAD

**Issue by fregante** [sindresorhus/eslint-plugin-unicorn#1706]
Occasionally it's easier to deal with IDs:
```jsx
const field = labelElement.htmlFor;
```
```jsx
return <input id={field}>;
```
```jsx
const element = document.getElementById(field)
```

It's kind of an exception, but this doesn't feel like an improvement:

```diff
const element = document.getElementById(field)
const element = document.querySelector('#' + field)
```

**Comment by fregante** [comment]
I had opened a duplicate issue with more examples 😅 

I'll copy it down here:

---

Related to:

- https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/prefer-query-selector.md

`querySelector` and `querySelectorAll` are great APIs, but are not necessarily "better" than the existing ones.

Both of these are fine, even though I'd also prefer the latter here:

```js
document.getElementsByClassname('class') ✅ 
document.querySelectorAll('.class')      ✅ 
```

However here qSA is making the code unnecessarily complex/noisy, to the point that qSA isn't any shorter either:

```js
document.getElementsByClassname(someClass) ✅
document.querySelectorAll('.' + someClass) 🥴
document.querySelectorAll(`.${someClass}`) 🥴
```

the same applies to `getElementById`, where this rule even lengthens the code:


```js
document.getElementById(someId)         ✅
document.querySelectorAll('.' + someId) 🤢
document.querySelectorAll(`.${someId}`) 🤢
```

Can we have an option to allow the """old""" APIs when they're used with existing variables?


```js
{
	"unicorn/prefer-query-selector": ["error", {
		"allowWithVariables": true
	}]
}
```

**Comment by sindresorhus** [comment]
I'm fine with an option as long as it's false by default.

**Comment by silverwind** [comment]
> querySelector and querySelectorAll are great APIs, but are not necessarily "better" than the existing ones.

I agree, they are useful but for cases like single class, single id etc, `querySelectorAll` is magnitudes slower then `getElementsByClassname` for example. In one benchmark I've seen it was 500 times slower in a DOM with 5000 elements.

**Comment by greypants** [comment]
I also agree - we ended up just disabling the rule for the following reasons:

- React `useId()` returns a selector like this: `:r8h:` (which IS a valid html ID value)
- The following works great: `document.getElementById(":r8h:")`
- But the rule will auto-fix that to: document.querySelector(`#:r8h:`) with will throw the following error:
   ```error
  SyntaxError: Failed to execute 'querySelector' on 'Document': '#:r8h:' is not a valid selector.
  ```
- To work around that, you have to escape it like this: `document.querySelector(CSS.escape(`#:r8h:`)) 
- The same issue will occur if an element id is set by a numeric id coming from the server.

It’s easier to just use document.getElementById(id)  and not have to remember to escape values or build a selector string with “#”

If this rule is going to *safely* autofix, it needs to treat `document.getElementById` like this:

```diff
- document.getElementById(id)
+ document.querySelector(`#${CSS.escape(id)}`)

**Comment by fregante** [comment]
Related:

- https://github.com/sindresorhus/eslint-plugin-unicorn/issues/2318

Given that with variables:

- you should use `CSS.escape()`
- you have to compose the selector by pretending `.` or `#`

This rule is a completely detrimental when used with variables. There isn't a single advantage in this case so it should ignore variables by default.
